apps/vmq_server/src/vmq_message_store.erl

```diff
@@ -46,50 +49,74 @@ stop() ->
     ok.
 
 write(SubscriberId, Msg) ->
+    write_with_retry(SubscriberId, Msg, ?NR_OF_RETRY).
+write_with_retry(SubscriberId, Msg, N) when N > 0 ->
     case vmq_util:timed_measurement({?MODULE, write}, vmq_plugin, only, [msg_store_write, [SubscriberId, Msg]]) of
-        {ok, N} ->
-            vmq_metrics:incr_stored_offline_messages(N);
+        {ok, Count} ->
+            vmq_metrics:incr_stored_offline_messages(Count);
+        {error, no_matching_hook_found} = ErrRes -> ErrRes;
         {error, Err} ->
             lager:error("Error: ~p", [Err]),
-            vmq_metrics:incr_msg_store_ops_error(write)
-    end,
-    ok.
+            vmq_metrics:incr_msg_store_ops_error(write),
+            timer:sleep(?RETRY_INTERVAL),
+            write_with_retry(SubscriberId, Msg, N-1)
+    end;
+write_with_retry(_SubscriberId, _Msg, _N) -> ok.
 
 read(SubscriberId, MsgRef) ->
+    read_with_retry(SubscriberId, MsgRef, ?NR_OF_RETRY).
+read_with_retry(SubscriberId, MsgRef, N) when N > 0 ->
```

Comment:
We should be measuring time taken to read for all attempts, not single attempts.